# HG changeset patch
# User Mike Miller <mtmiller@octave.org>
# Date 1464204119 25200
#      Wed May 25 12:21:59 2016 -0700
# Branch stable
# Node ID 1e81dd035639a8da963982f6a98e6e742b7eea22
# Parent  7496bd7265d9deae5826e3c036c0217bb55fbd57
Write integers with correct byte order on big-endian systems (bug #47434)

* oct-stream.cc (convert_data, octave_stream::write): Fix logic error in
determination of whether to swap bytes on big-endian systems.

diff --git a/libinterp/corefcn/oct-stream.cc b/libinterp/corefcn/oct-stream.cc
--- a/libinterp/corefcn/oct-stream.cc
+++ b/libinterp/corefcn/oct-stream.cc
@@ -3626,20 +3626,22 @@ public:
 template <class T>
 static bool
 convert_data (const T *data, void *conv_data, octave_idx_type n_elts,
               oct_data_conv::data_type output_type,
               oct_mach_info::float_format flt_fmt)
 {
   bool retval = true;
 
-  bool swap
-    = ((oct_mach_info::words_big_endian ()
-        && flt_fmt == oct_mach_info::flt_fmt_ieee_little_endian)
-       || flt_fmt == oct_mach_info::flt_fmt_ieee_big_endian);
+  bool swap = false;
+
+  if (oct_mach_info::words_big_endian ())
+    swap = (flt_fmt == oct_mach_info::flt_fmt_ieee_little_endian);
+  else
+    swap = (flt_fmt == oct_mach_info::flt_fmt_ieee_big_endian);
 
   bool do_float_conversion =  flt_fmt != oct_mach_info::float_format ();
 
   typedef typename ultimate_element_type<T>::type ult_elt_type;
 
   switch (output_type)
     {
     case oct_data_conv::dt_char:
@@ -3795,19 +3797,22 @@ octave_stream::skip_bytes (size_t skip)
 
 template <class T>
 octave_idx_type
 octave_stream::write (const Array<T>& data, octave_idx_type block_size,
                       oct_data_conv::data_type output_type,
                       octave_idx_type skip,
                       oct_mach_info::float_format flt_fmt)
 {
-  bool swap = ((oct_mach_info::words_big_endian ()
-                && flt_fmt == oct_mach_info::flt_fmt_ieee_little_endian)
-               || flt_fmt == oct_mach_info::flt_fmt_ieee_big_endian);
+  bool swap = false;
+
+  if (oct_mach_info::words_big_endian ())
+    swap = (flt_fmt == oct_mach_info::flt_fmt_ieee_little_endian);
+  else
+    swap = (flt_fmt == oct_mach_info::flt_fmt_ieee_big_endian);
 
   bool do_data_conversion = (swap || ! is_equivalent_type<T> (output_type)
                              || flt_fmt != oct_mach_info::float_format ());
 
   octave_idx_type nel = data.numel ();
 
   octave_idx_type chunk_size;
 
